contractcourt/channel_arbitrator.go

```diff
@@ -2393,39 +2410,40 @@ func (c *ChannelArbitrator) UpdateContractSignals(newSignals *ContractSignals) {
 	}
 }
 
-// contractUpdateSignal is a struct that carries the latest set of
-// ContractUpdate for a particular key. It also carries a done chan that should
-// be closed by the recipient.
-type contractUpdateSignal struct {
-	// newUpdate contains the latest ContractUpdate for a key.
-	newUpdate *ContractUpdate
-
-	// doneChan is an acknowledgement channel.
-	doneChan chan struct{}
+// notifyContractUpdate updates the ChannelArbitrator's unmerged mappings such
+// that it can later be merged with activeHTLCs when calling
+// checkLocalChainActions or sweepAnchors. These are the only two places that
+// activeHTLCs is used.
+func (c *ChannelArbitrator) notifyContractUpdate(upd *ContractUpdate) {
+	c.unmergedMtx.Lock()
+	defer c.unmergedMtx.Unlock()
+
+	// Update the mapping.
+	c.unmergedSet[upd.HtlcKey] = newHtlcSet(upd.Htlcs)
```

Comment:
Interesting, so we make this operation fully non-blocking vs potentially blocking on the channel write.

Alternatively, would just making `htlcUpdates` into a buffered channel resolve the issue?

Reply:
Yes, but there is more rationale here (https://github.com/lightningnetwork/lnd/issues/6312#issuecomment-1096850959)

Reply:
It seems unnecessary to introduce the concept of `unmergedSet`, rather, we could just do read/write to `activeHTLCs` inside a lock? To me the `unmergedSet` is equivalent to `activeHTLCs` but it's doing its operations inside a lock.

Reply:
good point didn't notice that

Reply:
I actually prefer the approach in this PR since you don't have to lock as long, which is why I didn't just use activeHTLCs with a mutex. Just using a mutex means having to hold it during `checkLocalChainActions` and `sweepAnchors` - the first func can query the preimage database, lookup invoices, etc. This way it's a lazy approach

Reply:
I don't think you need to hold it for the whole function, just like what the two functions now do essentially - `notifyContractUpdate` is a lock write, and `updateActiveHTLCs` is a lock read, from `activeHTLCs`'s perspective. We might as well just use `activeHTLCs` directly in those two.

Reply:
it has to be locked for the entire time for checkLocalChainActions since the map is passed to the func -- way around it is to copy the map every time which is what i wanted to avoid

Reply:
Cool I see. I was thinking about add a lock reader/writer around `activeHTLCs`, so access to `c.activeHTLCs` is forbidden and only `c.readHTLCs()` and `c.writeHTLCs(htlc)` is allowed. Anyway, I think either is fine.